ouroboros-consensus-protocol/src/unstable-protocol-testlib/Test/Ouroboros/Consensus/Protocol/Praos/Header.hs

```diff
@@ -0,0 +1,456 @@
+
+instance Json.FromJSON GeneratorContext where
+    parseJSON = Json.withObject "GeneratorContext" $ \obj -> do
+        praosSlotsPerKESPeriod <- obj .: "praosSlotsPerKESPeriod"
+        praosMaxKESEvo <- obj .: "praosMaxKESEvo"
+        rawKesSignKey <- obj .: "kesSignKey"
+        rawColdSignKey <- obj .: "coldSignKey"
+        rawVrfSignKey <- obj .: "vrfSignKey"
+        cborNonce <- obj .: "nonce"
+        ocertCounters <- obj .: "ocertCounters"
+        kesSignKey <- parseKesSignKey rawKesSignKey
+        coldSignKey <- parseColdSignKey rawColdSignKey
+        vrfSignKey <- parseVrfSignKey rawVrfSignKey
+        nonce <- parseNonce cborNonce
+        activeSlotCoeff <- mkActiveSlotCoeff <$> obj .: "activeSlotCoeff"
+        pure GeneratorContext{..}
+      where
+        parseNonce rawNonce =
+            case Base16.decode (encodeUtf8 rawNonce) of
+                Left _ -> pure NeutralNonce
+                Right nonceBytes -> Nonce <$> maybe (fail "invalid bytes for hash") pure (hashFromBytes nonceBytes)
+        parseColdSignKey rawKey = do
+            case Base16.decode (encodeUtf8 rawKey) of
+                Left err -> fail err
+                Right keyBytes ->
+                    case rawDeserialiseSignKeyDSIGN keyBytes of
+                        Nothing -> fail $ "Invalid cold key bytes: " <> show rawKey
+                        Just key -> pure key
+        parseKesSignKey rawKey = do
+            case Base16.decode (encodeUtf8 rawKey) of
+                Left err -> fail err
+                Right keyBytes ->
+                    case rawDeserialiseSignKeyKES keyBytes of
+                        Nothing -> fail $ "Invalid KES key bytes: " <> show rawKey
+                        Just key -> pure key
+        parseVrfSignKey rawKey = do
+            case Base16.decode (encodeUtf8 rawKey) of
+                Left err -> fail err
+                Right keyBytes ->
+                    case rawDeserialiseSignKeyVRF keyBytes of
+                        Nothing -> fail $ "Invalid VRF key bytes: " <> show rawKey
+                        Just key -> pure key
+
+genContext :: Gen GeneratorContext
+genContext = do
+    praosSlotsPerKESPeriod <- choose (100, 10000)
+    praosMaxKESEvo <- choose (10, 1000)
+    ocertCounter <- choose (10, 100)
+    kesSignKey <- newKESSigningKey <$> gen32Bytes
+    coldSignKey <- genKeyDSIGN . mkSeedFromBytes <$> gen32Bytes
+    vrfSignKey <- fst <$> newVRFSigningKey <$> gen32Bytes
+    nonce <- Nonce <$> genHash
+    let poolId = coerce $ hashKey $ VKey $ deriveVerKeyDSIGN coldSignKey
+        ocertCounters = Map.fromList [(poolId, ocertCounter)]
+    activeSlotCoeff <- genActiveSlotCoeff
+    pure $ GeneratorContext{..}
+
+genActiveSlotCoeff :: Gen ActiveSlotCoeff
+genActiveSlotCoeff = do
+    choose (1, 100) >>= \n -> pure $ activeSlotCoeff (n % 100)
+  where
+    activeSlotCoeff = mkActiveSlotCoeff . fromJust . boundRational @PositiveUnitInterval
```

Comment:
I usually understand it better if 0-arity `Gen`erators are just the `arbitrary` method of QuickCheck instead of a separate function, and then (>0)-arity `Gen`erators are clear that they cannot be generated on their own but they need some more data to be generated before.

Reply:
The `Arbitrary` instance is never used, as all the tests explicitly call generator functions. I like to "drive by needs" and only write the code that's necessary, for which I have written tests.

Furthermore, I think that in this case, we are more interested in specific scenarios, and I can imagine that we might want another generator for Context, eg. to highlight specific corner cases, or test different aspects related to `Header` construction and validation.